### ISSUE TITLE
SNS Subscriptions for Lambda functions

### ISSUE THREAD

**Issue by blakeblackshear** [hashicorp/terraform#2885]
I realize the lambda type is not listed in the documentation, but it seems to mostly work. It looks like the subscription is correctly added with the correct arn, but the lambda function never seems to get the notifications. However, if I create the same subscription manually in the Console, it starts working. The subscription information looks exactly the same when I create it manually as it did when created by terraform. Seems like something small is missing. After I make the manual change, `terraform plan` returns no changes.

**Comment by stack72** [comment]
@blakeblackshear so I have been through this exact issue this week. A lambda function needs to have an Event Source

That event source is handled easily in the UI *but* in the API it is handled very differently

Kinesis and DynamoDB will use the CreateEventSourceMapping API call

SNS and S3 are not streams. Therefore the lambda needs to have permissions granted to it to allow the subscription. The AppPermission API is a bit crap as it only has AddPermission and RevokePermission. 

http://docs.aws.amazon.com/cli/latest/reference/lambda/add-permission.html

Hope this helps a bit

P.

**Comment by blakeblackshear** [comment]
So can I handle this within the IAM role configuration on the lambda resource in terraform? Or do I need to manually call add-permission in a local provisioner?

**Comment by blakeblackshear** [comment]
Ok. I see. Further explaination here: https://mobile.awsblog.com/post/Tx1VE917Z8J4UDY/Invoking-AWS-Lambda-functions-via-Amazon-SNS

I guess this is different than giving permissions to the Lambda function through the IAM role.

**Comment by johnhamelink** [comment]
Hey there.

I'm coming up against the same issue as @blakeblackshear but I don't really understand the resoluton of this issue - did you ever get SNS -> Lambda integration working @blakeblackshear ?

**Comment by maxbeatty** [comment]
Thanks everyone for your questions and links. It helped me figure out a similar problem. @johnhamelink hope this is better late than never :smile_cat: 

I was able to invoke a Lambda function from an SNS notification by: 

1. creating an IAM role that assumes Lambda service
2. creating the SNS topic
3. creating a Lambda function
4. adding a permission to the Lambda function to be invoked by the SNS ARN (via `local-exec` provisioner)
5. creating a subscription to the SNS topic from the Lambda function 

Terraform configuration:
```
# 1. create IAM role assuming Lambda service
resource "aws_iam_role" "gfact-mc" {
  name = "gfact_lambda_exec"
  assume_role_policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Action": "sts:AssumeRole",
      "Principal": {
        "Service": "lambda.amazonaws.com"
      },
      "Effect": "Allow",
      "Sid": "Stmt1445981915064"
    }
  ]
}
EOF
}

# 2. create SNS topic
resource "aws_sns_topic" "example" {
  name = "example"
}

# 3. create Lambda function
resource "aws_lambda_function" "example" {
  depends_on = ["aws_sns_topic.example"]
  filename = "${var.lambda_example_build}"
  function_name = "${var.lambda_example}"
  handler = "index.handler"
  role = "${aws_iam_role.example.arn}"
  # 4. add permission to lambda function with local provisioner
  provisioner "local-exec" {
    command = "AWS_ACCESS_KEY_ID=${var.access_key} AWS_SECRET_ACCESS_KEY=${var.secret_key} TF_REGION=${var.region} TF_LAMBDA_NAME=${var.lambda_example} TF_SNS_ARN=${aws_sns_topic.example.id} node provisioners/link-sns-to-lambda.js"
  }
}

# 5. subscribe Lambda function to SNS topic
resource "aws_sns_topic_subscription" "example" {
  depends_on = ["aws_lambda_function.example"]
  topic_arn = "${aws_sns_topic.example.arn}"
  protocol = "lambda"
  endpoint = "${aws_lambda_function.example.arn}"
}
```

Node.js `link-sns-to-lambda.js` (`aws-sdk` requires `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`)
```js
var aws = require('aws-sdk');

var lambda = new aws.Lambda({
  apiVersion: '2015-03-31',
  region: process.env.TF_REGION,
});

lambda.addPermission({
  FunctionName: process.env.TF_LAMBDA_NAME,
  StatementId: Date.now().toString(),
  Action: 'lambda:InvokeFunction',
  Principal: 'sns.amazonaws.com',
  SourceArn: process.env.TF_SNS_ARN,
}, function callback(err, data) {
  if (err) {
    throw err;
  } else {
    console.log('added permission to lambda', process.env.TF_LAMBDA_NAME);
    console.dir(data);
  }
});
```

**Comment by stack72** [comment]
@blakeblackshear have you had a chance to see if this would work for you? I have since done the same thing as above and it seems to be the only way to do it right now

Thoughts?

**Comment by blakeblackshear** [comment]
Haven't had a chance to try it, but it looks good to me as long as the call
in local-exec is idempotent. I would probably opt for using the aws-cli
instead of a node script if possible.

@blakeblackshear <https://github.com/blakeblackshear> have you had a chance
to see if this would work for you? I have since done the same thing as
above and it seems to be the only way to do it right now

Thoughts?

—
Reply to this email directly or view it on GitHub
<https://github.com/hashicorp/terraform/issues/2885#issuecomment-152672267>.

**Comment by netpulse-ci-server** [comment]
**aws_lambda_event_source_mapping** resource, added in 0.6.8 release doesn't seem to work with SNS either, thus, this issue still persists.

```
...
InvalidParameterValueException: Unrecognized event source, must be kinesis or dynamodb stream. Arn: Unsupported stream arn: arn:aws:sns...
...
```

**Comment by smotovilovets** [comment]
**aws_lambda_event_source_mapping** resource, added in 0.6.8, doesn't seem to work with SNS either, thus, this issue persists.

```
...
Unrecognized event source, must be kinesis or dynamodb stream. Arn: Unsupported stream arn: arn:aws:sns:...
...
```

**Comment by sp52005** [comment]
@blakeblackshear: 
  I have the same issue before. After looking the hints above, here is my solution with only two aws cli commands (source-arn is a key attribute in the first add-permission command.) as following.

    aws lambda add-permission --function-name <lambda name> --statement-id sns-invoke --action lambda:invokeFunction --principal sns.amazonaws.com --region <region> --source-arn <topic arn>
    aws sns subscribe --region <region> --topic-arn <topic arn> --protocol lambda --notification-endpoint <lambda arn>

**Comment by jtollman** [comment]
@sp52005 
I literally just came to the same conclusion as you, tried it, and it STILL doesn't work.  Does that work for you?  I can certainly get it working via the Lambda console, but not via CLI calls.  Further, when I add the permissions, while I can see them via aws lambda get-policy, it doesn't show up in the Lambda console like the others I have enabled using the console.  I have been banging my head against this for too long, might me time for a support call.

**Comment by jtollman** [comment]
Scratch that, @sp52005 solution is correct.  My ARN was malformed (was using the response ARN from the SNS subscription which has an event identifier tagged on to the end).

**Comment by stack72** [comment]
Hi, this should now be fixed by #4826
